docs/advanced-usage.rst

```diff
@@ -0,0 +1,176 @@
+
+    >>> r = http.request(
+    ...     'GET',
+    ...     'http://httpbin.org/bytes/1024',
+    ...     preload_content=False)
+    >>> r.read(4)
+    b'\x88\x1f\x8b\xe5'
+
+Calls to :meth:`~response.HTTPResponse.read()` will block until more response
+data is available. 
+
+    >>> import io
+    >>> reader = io.BufferedReader(r, 8)
+    >>> reader.read(4)
+    >>> r.release_conn()
+
+You can use this file-like object to do things like decode the content using
+:mod:`codecs`::
+
+    >>> import codecs
+    >>> reader = codecs.getreader('utf-8')
+    >>> r = http.request(
+    ...     'GET',
+    ...     'http://httpbin.org/ip',
+    ...     preload_content=False)
+    >>> json.load(reader(r))
+    {'origin': '127.0.0.1'}
+    >>> r.release_conn()
+
+.. _proxies:
+
+Proxies
+-------
+
+You can use :class:`~poolmanager.ProxyManager` to tunnel requests through an
+HTTP proxy::
+
+    >>> proxy = urllib3.ProxyManager('http://localhost:3128/')
+    >>> proxy.request('GET', 'http://google.com/')
+
+The usage of :class:`~poolmanager.ProxyManager` is the same as
+:class:`~poolmanager.PoolManager`.
+
+You can use :class:`~contrib.socks.SOCKSProxyManager` to connect to SOCKS4 or
+SOCKS5 proxies. In order to use SOCKS proxies you will need to install
+`PySocks <https://pypi.python.org/pypi/PySocks>`_ or install urllib3 with the
+``socks`` extra::
+
+    pip install urllib3[socks]
+
+Once PySocks is installed, you can use
+:class:`~contrib.socks.SOCKSProxyManager`::
+
+    >>> from urllib3.contrib.socks import SOCKSProxyManager
+    >>> proxy = SOCKSProxyManager('socks5://localhost:8889/')
+    >>> proxy.request('GET', 'http://google.com/')
+
+.. _ssl_mac:
+
+Certificate validation and Mac OS X
+-----------------------------------
+
+Apple-provided Python and OpenSSL libraries contain a patches that make them
+automatically check the system keychain's certificates. This can be
+surprising if you specify custom certificates and see requests unexpectedly
+succeed. See this
+`article <https://hynek.me/articles/apple-openssl-verification-surprises/>`_
+for more information.
+
+.. _ssl_warnings:
+
+SSL Warnings
+------------
+
+urllib3 will issue several different warnings based on the level of certificate
+verification support. These warning indicate particular situations and can
+resolved in different ways.
+
+* :class:`~exceptions.InsecureRequestWarning`
+    This happens when an request is made to an HTTPS URL without certificate
+    verification enabled. Follow the :ref:`certificate verification <ssl>`
+    guide to resolve this warning.
+* :class:`~exceptions.InsecurePlatformWarning`
+    This happens on Python 2 platforms that have an outdated :mod:`ssl` module.
+    These older :mod:`ssl` modules can cause some insecure requests to succeed
+    where they should fail and secure requests to fail where they should
+    succeed. Follow the :ref:`pyOpenSSL <ssl_py2>` guide to resolve this
+    warning.
```

Comment:
For this and the SNI warning, do we want to also include "upgrade to a newer version of Python if possible" as an option? @jonparrott, don't make the change - I'd like opinions from @shazow and/or @Lukasa.